remotes/push.go

```diff
@@ -235,3 +245,75 @@ func pushBundleConfigDescriptor(ctx context.Context, name string, resolver remot
 	}
 	return descriptor, nil
 }
+
+func pushUnderTag(ctx context.Context, imageClient client.ImageAPIClient, image *bundle.BaseImage, targetRef reference.Named) error {
+	ref := image.Digest
+	if ref == "" {
+		ref = image.Image
+	}
+
+	if err := imageClient.ImageTag(ctx, ref, targetRef.String()); err != nil {
+		return err
+	}
+
+	err := pushTaggedImage(ctx, imageClient, targetRef)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func pushTaggedImage(ctx context.Context, imageClient client.ImageAPIClient, targetRef reference.Named) error {
+	repoInfo, err := registry.ParseRepositoryInfo(targetRef)
+	if err != nil {
+		return err
+	}
+
+	authConfig := resolveAuthConfig(repoInfo.Index)
+	encodedAuth, err := encodeAuthToBase64(authConfig)
+	if err != nil {
+		return err
+	}
+
+	reader, err := imageClient.ImagePush(ctx, targetRef.String(), types.ImagePushOptions{
+		RegistryAuth: encodedAuth,
+	})
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+	if err := jsonmessage.DisplayJSONMessagesStream(reader, ioutil.Discard, 0, false, nil); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func encodeAuthToBase64(authConfig types.AuthConfig) (string, error) {
+	buf, err := json.Marshal(authConfig)
+	if err != nil {
+		return "", err
+	}
+	return base64.URLEncoding.EncodeToString(buf), nil
+}
+
+func resolveAuthConfig(index *registrytypes.IndexInfo) types.AuthConfig {
```

Comment:
Also, this is going to be really painful to vendor if you don't have the exact revisions...


```
# github.com/engineerd/signy/vendor/github.com/docker/cnab-to-oci/remotes
vendor/github.com/docker/cnab-to-oci/remotes/push.go:318:2: cannot use authConfig (type "github.com/engineerd/signy/vendor/github.com/docker/cli/cli/config/types".AuthConfig) as type "github.com/engineerd/signy/vendor/github.com/docker/docker/api/types".AuthConfig in return argument
```

Reply:
Right, thank you for catching that @radu-matei 👍 
We need to be more strict on our dependencies 😅

Reply:
I've pushed a new bunch of commits to improve dependencies of `cnab-to-oci`. I'm not sure this is the right way of doing it.

I mainly upgraded the `cnab-go` to the latest tag and let it to get `docker/cli` and `docker/docker` (`cli` is the same as before for `cnab-to-oci`, `docker` is a little newer)

From what I saw it will not fix this issue, because the `AuthConfig` type has been introduced in `docker/cli` more recently.
I supposed the better way to fix this issue should be to upgrade both `cnab-go` and `cnab-to-oci` to latest releases of `docker/cli` and `docker/docker` but at least for `cnab-to-oci` it requires some more work.

Dependency upgrades are in dedicated commits so we can easily choose to not to do it.

Reply:
@radu-matei can you check it works on your side?

Reply:
About cnab-go, we are working with @silvin-lubecki to remove the dep on the cli, we only need the api client. Will post something there soon (next week).

Reply:
Opened https://github.com/deislabs/cnab-go/issues/146 to track this in `cnab-go`.